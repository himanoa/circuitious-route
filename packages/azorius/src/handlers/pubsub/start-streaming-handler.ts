import { StartStreamingEvent } from "izett";
import { Client, VoiceChannel, Channel } from "discord.js";
import { Logger } from "log4js";

export type Dependencies = {
  discordClient: Client;
  logger: Logger;
};

const isVoiceChannel = (c: Channel): c is VoiceChannel => c.type === "voice";

export const startStreamingHandler = (deps: Dependencies) => (
  event: StartStreamingEvent
) => {
  const { discordId } = event.payload;
  const voiceChannels = deps.discordClient.channels.cache.filter(
    (c) => c.type === "voice"
  );
  const streamerUserJoinedVoiceChannel = voiceChannels.find((v) => {
    if (isVoiceChannel(v)) {
      return Array.from(v.members.keys()).includes(discordId);
    }
    return false;
  });

  if (
    streamerUserJoinedVoiceChannel &&
    isVoiceChannel(streamerUserJoinedVoiceChannel)
  ) {
    return streamerUserJoinedVoiceChannel.join();
  }
  deps.logger.info(
    "ボイスチャンネルに配信開始したユーザーが存在しませんでした"
  );
};
