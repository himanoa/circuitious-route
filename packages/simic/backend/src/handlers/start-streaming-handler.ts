import * as Express from "express";
import { Events } from "izett";
import { TokenNotFoundError } from "../error";

export const startStreamingHandler = (deps: {
  publishEvent: (event: Events) => Promise<void>;
  verify: (token: string) => { discordId: string };
}) => async (req: Express.Request, res: Express.Response) => {
  const tokenWithType = req.get("Authorization");
  if (!tokenWithType) {
    throw new TokenNotFoundError("Token is not found");
  }

  try {
    const [, token] = tokenWithType.split(" ");
    const { discordId } = deps.verify(token);
    await deps.publishEvent({
      type: "start-streaming",
      payload: {
        discordId,
      },
    });
    res.status(200).json({});
  } catch (err) {
    if (err instanceof TokenNotFoundError) {
      res.status(401).json({ error: err });
    }
    console.dir(err);
    res.status(500).json({ error: err });
  }
};
