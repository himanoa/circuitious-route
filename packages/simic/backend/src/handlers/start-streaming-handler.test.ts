import { startStreamingHandler } from "./start-streaming-handler";
import * as Response from "jest-express/lib/response";
import * as Request from "jest-express/lib/request";

describe("start-streaming-handler", () => {
  describe("when verify successful", () => {
    const target = startStreamingHandler({
      publishEvent: () => Promise.resolve(),
      verify: () => ({ discordId: "12123123".toString() }),
    });

    it("should be failed", async () => {
      const request = new Request.Request("https://example.com");
      request.setHeaders("Authorization", "Barer InvalidToken");
      const response = new Response.Response();
      await target(request as any, response as any);
      expect(response.status).toBeCalledWith(200);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("when verify failed", () => {
    const target = startStreamingHandler({
      publishEvent: () => Promise.resolve(),
      verify: () => {
        throw new Error("doo");
      },
    });

    it("should be failed", async () => {
      const request = new Request.Request("https://example.com");
      request.setHeaders("Authorization", "Barer InvalidToken");
      const response = new Response.Response();
      await target(request as any, response as any);
      expect(response.status).toBeCalledWith(500);
      expect(response.body).toMatchSnapshot();
    });
  });
});
