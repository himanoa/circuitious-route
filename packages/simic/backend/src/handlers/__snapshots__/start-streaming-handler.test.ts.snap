// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start-streaming-handler when verify failed should be failed 1`] = `
Object {
  "error": [Error: doo],
}
`;

exports[`start-streaming-handler when verify successful should be failed 1`] = `Object {}`;
